core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -140,92 +144,63 @@ String getPreferredIp(List<String> preferredTypes) {
 
   /**
    * Attempts to force a new refresh of the instance data. May fail if called too frequently or if a
-   * new refresh is already in progress. If successful, other methods may block until refresh has
+   * new refresh is already in progress. If successful, other methods will block until refresh has
    * been completed.
    *
-   * @return {@code true} if successfully scheduled, or {@code false} if it is called too
-   *     frequently.
+   * @return {@code true} if successfully scheduled, or {@code false} otherwise.
    */
   boolean forceRefresh() {
     if (!forcedRenewRateLimiter.tryAcquire()) {
-      return false; // Forced refreshing too often
+      return false;
     }
     synchronized (instanceDataGuard) {
-      // If a scheduled Refresh already exists, cancel if if not immediately imminent
-      if (nextInstanceData == null || nextInstanceData.cancel(false)) {
-        // If canceled successfully, schedule a replacement immediately
-        nextInstanceData = null;
-        scheduleRefresh(0, TimeUnit.MINUTES);
+      // If a scheduled refresh exists and can't be canceled, fail the operation
+      if (nextInstanceData != null && !nextInstanceData.cancel(false)) {
```

Comment:
There's still the case where someone calls `forceRefresh()` while the initial future is pending.

1. The constructor calls `forceRefresh()`. `nextInstanceData` is `null` until that refresh completes.
3. Before that refresh completes, someone calls `forceRefresh()`.
4. Since `nextInstanceData` is still `null`, the `if` condition is false, so `currentInstanceData` is replaced by a new incomplete future.
5. The first operation completes, and replaces `currentInstanceData` with the first completed future.
6. The second operation completes, and replaces `currentInstanceData` with the second completed future.

That will be okay for the users; it shouldn't mean any extra delays. But it does mean more than one pending operation at once.

Reply:
An alternative is to initialize `nextInstanceData` with a completed future, maybe `Futures.immediateFuture(null)`. Then it'll never be `null`, and in that state won't be cancellable.

Reply:
Would `Futures.immediateCancelledFuture()` make more sense than `Futures.immediateFuture(null)`?